PyDynamic/model_estimation/fit_filter.py

```diff
@@ -87,92 +86,207 @@ def _fitIIR(
-        print("Final sum of squares = %e" % sos)
-        tau = 0
-        return b, a, tau
-
-    if np.count_nonzero(np.abs(np.roots(a)) > 1) > 0:
-        stable = False
-    else:
-        stable = True
-
-    maxiter = 50
-
-    astab = mapinside(a)
-    run = 1
-
-    while stable is not True and run < maxiter:
-        g1 = grpdelay(b, a, Fs)[0]
-        g2 = grpdelay(b, astab, Fs)[0]
-        tau = np.ceil(tau + np.median(g2 - g1))
-
-        b, a = _fitIIR(Hvals, tau, w, E, Na, Nb, inv=False)
-        if np.count_nonzero(np.abs(np.roots(a)) > 1) > 0:
-            astab = mapinside(a)
+    as_and_bs = np.empty((mc_runs, Nb + Na + 1))
+    taus = np.zeros((mc_runs,), dtype=int)
+    stab_iters = np.zeros((mc_runs,), dtype=int)
+
+    # Conduct the Monte Carlo runs or in case we did not have uncertainties execute
+    # just once the actual algorithm.
+    for mc_run in range(mc_runs):
+        # Conduct actual fit.
+        b_i, a_i = _fitIIR(Hvals, tau, w, E, Na, Nb, inv=inv)
+
+        # Determine if the computed filter already is stable.
+        unstable = not isstable(b=b_i, a=a_i, ftype="digital")
+
+        # In case the user specified not to check for stability,
+        # we skip the rest of the current Monte Carlo run and inform the user.
+        if justFit:
+            if verbose:
+                sos = np.sum(np.abs((dsp.freqz(b_i, a_i, w)[1] - Hvals) ** 2))
+                print(
+                    f"LSIIR: Fitting "
+                    f"{'' if UHvals is None else f'for MC run {mc_run} '}"
+                    f"finished. No stabilization requested"
+                    f". {warning_unstable if unstable else ''} "
+                    f"Final sum of squares = {sos}"
+                )
+            # Since no stabilization was done, we return tau = 0 regardless of the
+            # desired initial estimate of the time delay provided for the stabilization.
+            taus[mc_run] = 0
+            continue
         else:
-            stable = True
-        run = run + 1
-
-    if np.count_nonzero(np.abs(np.roots(a)) > 1) > 0:
-        print("Caution: The algorithm did NOT result in a stable IIR filter!")
-        print(
-            "Maybe try again with a higher value of tau0 or a higher filter "
-            "order?")
+            # Otherwise we initialize the time delay for the current Monte Carlo run.
+            taus[mc_run] = tau
+
+        # Initialize counter which we use to report about required iteration count.
+        current_stab_iter = 0
+        # Stabilize filter coefficients with a maximum number of iterations.
+        while unstable and current_stab_iter < max_stab_iter:
+            # Compute appropriate time delay for the stabilization of the filter.
+            a_stab = mapinside(a_i)
+            g_1 = grpdelay(b_i, a_i, Fs)[0]
+            g_2 = grpdelay(b_i, a_stab, Fs)[0]
+            taus[mc_run] += np.ceil(np.median(g_2 - g_1))
```

Comment:
For the final result we should store the maximum tau, for which we still received a stable filter. In each subsequent loop, we could then check for stability of the fit itself, then for  stability after the previously determined delay or further iterate.

Reply:
We introduced this today.